### PR TITLE
parquet Statistics - deprecate `has_*` APIs and add `_opt` functions that return `Option<T>`

### LINKED ISSUE #6093: Remove `Statistics::has_min_max_set` and `ValueStatistics::has_min_max_set` and use `Option` instead
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

The fact that Paruqet metadata statistics requires checking before access is confusing and error prone, for example https://github.com/apache/arrow-rs/pull/6092

**Describe the solution you'd like**

Change the API so it doesn't panic if used incorrectly

**Describe alternatives you've considered**
I propose removing these methods:
* https://docs.rs/parquet/latest/parquet/file/statistics/enum.Statistics.html#method.has_min_max_set
* https://docs.rs/parquet/latest/parquet/file/statistics/struct.ValueStatistics.html#method.has_min_max_set

And changing the corresponding accessors like `min()` and `max()` to return `Option<&T>` rather than `&T`

Today:
```rust
impl Statistics { 
...
  // panics if `has_min_max_set` is returns false
  pub fn min(&self) -> &T
...
}
```

Proposal:
```rust
impl Statistics { 
...
  // Returns None if statistics not set
  pub fn min(&self) -> Option<&T>
...
}
```

**Additional context**

### LINKED ISSUE #6215: Parquet Statistics null_count does not distinguish between `0` and not specified
**Describe the bug**

Please close if this is intended: currently the [null_count](https://github.com/apache/arrow-rs/blob/master/parquet/src/file/statistics.rs#L129-L130) in `Statistics` defaults to 0 and the [has_nulls()](https://github.com/apache/arrow-rs/blob/master/parquet/src/file/statistics.rs#L393) is based on `null_count > 0`, leading to ambiguity when `null_count` equals `0`: either all values are non-null, or just that the null count stat is missing.

**To Reproduce**
n/a

**Expected behavior**

it feels reasonable to distinguish `no null` and `null_count not set`.

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by Michael-J-Ward**
# Which issue does this PR close?

Closes #6093
Closes #6215

# Rationale for this change
 
From the feature request for `min` and `max`
> The fact that Parquet metadata statistics requires checking before access is > confusing and error prone, for example https://github.com/apache/arrow-rs/pull/6092

From the feature request for `null_count`
> and the [has_nulls()](https://github.com/apache/arrow-rs/blob/master/parquet/src/file/statistics.rs#L393) is based on null_count > 0, leading to ambiguity when null_count equals 0: either all values are non-null, or just that the null count stat is missing.

# What changes are included in this PR?

## APIs added

- `Statistics::{min,max}_bytes_opt`
- `ValueStatistics::{min,max}_bytes_opt`
- `ValueStatistics::{min,max}_opt`
- `Statistics::null_count_opt`
- `ValueStatistics::null_count_opt`

## APIs deprecated
- `Statistics::has_nulls`
- `Statistics::null_count`
- `ValueStatistics::null_count`
- `Statistics::has_min_max_set` 
- `ValueStatistic::has_min_max_set`
- `Statistics::{min,max}_bytes`
- `ValueStatistics::{min,max}_bytes`
- `ValueStatistics::{min,max}`

# Are there any user-facing changes?

Yes. All of the above changes are changes to the public APi.

Additionally, `null_count = 0` is now written to page statistics instead of being treated as None.

### CODE REVIEW

File: parquet/src/arrow/arrow_reader/statistics.rs [hunk 1]
```diff
@@ -131,24 +131,64 @@ macro_rules! make_stats_iterator {
...
+);
+make_stats_iterator!(
+    MaxDoubleStatsIterator,
+    max_unchecked,
+    ParquetStatistics::Double,
+    f64
+);
 make_stats_iterator!(
     MinByteArrayStatsIterator,
     min_bytes,
```
Comment (Michael-J-Ward) [inline parquet/src/arrow/arrow_reader/statistics.rs:194]: If we change `Statistics::{min,max}_bytes` to return `Option<&[u8]>`, The use of `*_unchecked` methods can be switched to the new imps and just use `unwrap`

File: parquet/src/column/writer/mod.rs [hunk 2]
```diff
@@ -769,8 +769,8 @@ impl<'a, E: ColumnValueEncoder> GenericColumnWriter<'a, E> {
                 }
                 Some(stat) => {
                     // Check if min/max are still ascending/descending across pages
-                    let new_min = stat.min();
-                    let new_max = stat.max();
+                    let new_min = stat.min_unchecked();
```
Comment (Michael-J-Ward) [inline parquet/src/column/writer/mod.rs:772]: Previously, this method did not check `has_minx_max_set` before calling `stat.min()`.

Should it have or is it guaranteed to already be set by this point?

File: parquet/src/file/statistics.rs [hunk 3]
```diff
@@ -385,18 +380,13 @@ impl Statistics {
...
 
-    /// Returns `true` if statistics collected any null values, `false` otherwise.
-    pub fn has_nulls(&self) -> bool {
-        self.null_count() > 0
-    }
-
     /// Returns `true` if min value and max value are set.
     /// Normally both min/max values will be set to `Some(value)` or `None`.
-    pub fn has_min_max_set(&self) -> bool {
+    pub(crate) fn has_min_max_set(&self) -> bool {
```
Comment (Michael-J-Ward) [inline parquet/src/file/statistics.rs:389]: This method is used both in tests for assertions and for logical flow in source code, so I kept it as internal to this crate.

File: parquet/src/file/statistics.rs [hunk 4]
```diff
@@ -601,6 +601,7 @@ impl<T: ParquetValueType> ValueStatistics<T> {
     /// using signed comparison. This resulted in an undefined ordering for unsigned
     /// quantities, such as booleans and unsigned integers.
     ///
+    // TODO: I don't see `min_value` or `max_value` in this struct. Should these be something else?
```
Comment (Michael-J-Ward) [inline parquet/src/file/statistics.rs:604]: And follow-up in case there should be other fields, should `min_bytes` referr to the deprecated `min` or to the new `min_value`

**Review by Michael-J-Ward — COMMENTED** [review]
Lastly, I'd like to call attention to the change in memory size by switching from `null_count: u64` to `null_count: Option<u64>`.

I updated the memory size and layout tests to pass, but I'm unsure if those values were "sacred".

**Review by crepererum — APPROVED** [review]
I've also ran into this weird API before (which I think is inspired by certain C++ interfaces) and I think this new API is better and more "Rusty".

**Comment by crepererum** [comment]
I'm wondering if we need an API deprecation here or not. If we need one, we would need to use an approach like this:

- `has_...` and `min`/`max`/....: add deprecation note
- add new `Option` APIs: `min_opt`/`max_opt`/...

That would be the smoothest path. However, the API breakage is rather simple and easy to fix. @alamb WDYT?

**Comment by Michael-J-Ward** [comment]
@crepererum - I'm happy to go either way, just let me know.

I went with the breaking change because I thought removing such a foot-gun is suitable for a breaking-change / major-release upgrade.  Also, that's what the GH issues requested.

File: parquet/tests/arrow_writer_layout.rs [hunk 5]
```diff
@@ -189,7 +189,7 @@ fn test_primitive() {
                     pages: (0..8)
                         .map(|_| Page {
                             rows: 250,
-                            page_header_size: 36,
+                            page_header_size: 38,
```
Comment (alamb) [inline parquet/tests/arrow_writer_layout.rs:192]: why are these changes needed?

File: parquet/src/file/statistics.rs [hunk 6]
```diff
@@ -385,18 +380,13 @@ impl Statistics {
 
     /// Returns number of null values for the column.
     /// Note that this includes all nulls when column is part of the complex type.
```
Comment (alamb) [inline parquet/src/file/statistics.rs:382]:
```suggestion
    /// Returns number of null values for the column, if known.
    /// Note that this includes all nulls when column is part of the complex type.
```

File: parquet/src/file/statistics.rs [hunk 7]
```diff
@@ -538,37 +528,47 @@ impl<T: ParquetValueType> ValueStatistics<T> {
     ///
     /// Panics if min value is not set, e.g. all values are `null`.
     /// Use `has_min_max_set` method to check that.
-    pub fn min(&self) -> &T {
+    pub(crate) fn min_unchecked(&self) -> &T {
         self.min.as_ref().unwrap()
     }
 
+    /// Returns min value of the statistics.
```
Comment (alamb) [inline parquet/src/file/statistics.rs:535]:
```suggestion
    /// Returns min value of the statistics, if known.
```

File: parquet/src/column/writer/mod.rs (hunk 2 above)
Comment (alamb) [reply parquet/src/column/writer/mod.rs:772]: I think it is guaranteed to be set at this point (this is the code that is responsible for writing the statistics)

File: parquet/src/file/statistics.rs (hunk 4 above)
Comment (alamb) [reply parquet/src/file/statistics.rs:604]: I don't think we should worry about making a nicer API for the older versions of parquet. We can do so if/when someone has such a file and a real usecase

**Review by alamb — COMMENTED** [review]
Thank you @Michael-J-Ward this is epic. Thank you @crepererum for the review

In my opinion we should make all the APIs  consistent (aka change `max_bytes` to return `Option<&[u8]>` etc

Responding to @crepererum 's comment in https://github.com/apache/arrow-rs/pull/6216#issuecomment-2283809587

> has_... and min/max/....: add deprecation note
> add new Option APIs: min_opt/max_opt/...
> That would be the smoothest path. However, the API breakage is rather simple and easy to fix. @alamb WDYT?

I would also be ok with either path as long as the API is consistent (e.g. `min_bytes` should do the same thing as `min`)

I would personally prefer erring on the "nicer  experience for users of this crate" and thus go the backwards compatible route:

1. Leave the existing functions as is but mark them deprecated
2. Add new functions like `min_opt()`, `max_opt()`, etc that return `Option<..>`

That means users will be told what is going on with the deprecated APIs and how to fix it, even though the eventual API may not be as concise.

File: parquet/tests/arrow_writer_layout.rs (hunk 5 above)
Comment (Michael-J-Ward) [reply parquet/tests/arrow_writer_layout.rs:192]: These sizes changed upon switching `ValueStatistic::null_count` from `u64` to `Option<u64>`.

Although I'd expect such a change to require an extra bit, I was still concerned that these sizes might be set by the parquet spec, and so called it out.

https://github.com/apache/arrow-rs/pull/6216#pullrequestreview-2229709079

File: parquet/src/file/statistics.rs [hunk 8]
```diff
@@ -246,11 +245,7 @@ pub fn to_thrift(stats: Option<&Statistics>) -> Option<TStatistics> {
     let mut thrift_stats = TStatistics {
         max: None,
         min: None,
-        null_count: if stats.has_nulls() {
-            Some(stats.null_count() as i64)
-        } else {
-            None
-        },
+        null_count: stats.null_count_opt().map(|value| value as i64),
```
Comment (Michael-J-Ward) [inline parquet/src/file/statistics.rs:248]: @alamb - this is why the `arrow_writer_layout` tests changed.

The previous API treated `null_count = 0` as None.

The new API treats `null_count = 0` as `Some(0)`.

I believe the new behavior is what is desired, but can easily revert to the old behavior with:

```
        null_count: stats.null_count_opt().map(|value| value as i64).filter(|&x| x > 0),
```

File: parquet/tests/arrow_writer_layout.rs (hunk 5 above)
Comment (Michael-J-Ward) [reply parquet/tests/arrow_writer_layout.rs:192]: Scratch the above. The answer is in the linked comment.

When I initially converted `null_count` to `Option<64>`, the *first* test I updated `test_memory_size`, and incorrectly assume the rest of the layout tests were downstream of that one.

https://github.com/apache/arrow-rs/pull/6216#discussion_r1716983500

**Comment by Michael-J-Ward** [comment]
@alamb and @crepererum - this PR is ready for another look.

**Review by crepererum — APPROVED** [review]

File: parquet/src/file/statistics.rs (hunk 8 above)
Comment (alamb) [reply parquet/src/file/statistics.rs:248]: Update -- I think this code is related to reading the statistics that were in the parquet file rather than how they are written. 

It seems like previously this code just set the statistics count to zero https://github.com/apache/arrow-rs/issues/6256 tracks this

Comment (alamb) [reply parquet/src/file/statistics.rs:248]: I agree the new behavior is desired, but I think it changes what values are written to parquet files (specifically the parquet metadata will now have the thrift equivalent of `Some(0)` rather than the equivalent of `None`. I filed https://github.com/apache/arrow-rs/issues/6256 to track

As this PR is already quite large, I think we should split it into two parts:
1. The API changes
2. The change for writing the metadata

I plan to update this PR to revert the changes to the metadata writing, and will then make a follow on PR to discuss / propose changing the statistics that are written to the file

**Review by alamb — COMMENTED** [review]
Thank you so much for this PR @Michael-J-Ward  -- this improvement is a very long time coming. 

I plan to split it into two parts -- I'll leave this PR with the API change and make a new PR with the change to the files that are written

**Comment by alamb** [comment]
I pushed [d4e650](https://github.com/apache/arrow-rs/pull/6216/commits/7d4e65076f9b1ea61f6931a670436cc08f4f629a) to revert the writer behavior changes and added some comments

**Comment by alamb** [comment]
I am going to merge this PR as the APIs code has been reviewed and the content of what is read/written to statistics is the same as on master. I will open follow on PRs for discussion and am happy to make other changes if desired

**Review by alamb — APPROVED** [review]
Thanks again @Michael-J-Ward -- epic

**Comment by alamb** [comment]
A small follow up from this PR https://github.com/apache/arrow-rs/pull/6259
